## 1. The oldest businesses in the world
<p>This is Staffelter Hof Winery, Germany's oldest business, which was established in 862 under the Carolingian dynasty. It has continued to serve customers through dramatic changes in Europe such as the Holy Roman Empire, the Ottoman Empire, and both world wars. What characteristics enable a business to stand the test of time? Image credit: <a href="https://commons.wikimedia.org/wiki/File:MKn_Staffelter_Hof.jpg">Martin Kraft</a>
<img src="https://assets.datacamp.com/production/project_1383/./img/MKn_Staffelter_Hof.jpeg" alt="The entrance to Staffelter Hof Winery, a German winery established in 862." width="400"></p>
<p>To help answer this question, BusinessFinancing.co.uk <a href="https://businessfinancing.co.uk/the-oldest-company-in-almost-every-country">researched</a> the oldest company that is still in business in almost every country and compiled the results into a dataset. Let's explore this work to to better understand these historic businesses. Our datasets, which are all located in the <code>datasets</code> directory, contain the following information: </p>
<h3 id="businessesandnew_businesses"><code>businesses</code> and <code>new_businesses</code></h3>
<table>
<thead>
<tr>
<th style="text-align:left;">column</th>
<th>type</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;"><code>business</code></td>
<td>varchar</td>
<td>Name of the business.</td>
</tr>
<tr>
<td style="text-align:left;"><code>year_founded</code></td>
<td>int</td>
<td>Year the business was founded.</td>
</tr>
<tr>
<td style="text-align:left;"><code>category_code</code></td>
<td>varchar</td>
<td>Code for the category of the business.</td>
</tr>
<tr>
<td style="text-align:left;"><code>country_code</code></td>
<td>char</td>
<td>ISO 3166-1 3-letter country code.</td>
</tr>
</tbody>
</table>
<h3 id="countries"><code>countries</code></h3>
<table>
<thead>
<tr>
<th style="text-align:left;">column</th>
<th>type</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;"><code>country_code</code></td>
<td>varchar</td>
<td>ISO 3166-1 3-letter country code.</td>
</tr>
<tr>
<td style="text-align:left;"><code>country</code></td>
<td>varchar</td>
<td>Name of the country.</td>
</tr>
<tr>
<td style="text-align:left;"><code>continent</code></td>
<td>varchar</td>
<td>Name of the continent that the country exists in.</td>
</tr>
</tbody>
</table>
<h3 id="categories"><code>categories</code></h3>
<table>
<thead>
<tr>
<th style="text-align:left;">column</th>
<th>type</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;"><code>category_code</code></td>
<td>varchar</td>
<td>Code for the category of the business.</td>
</tr>
<tr>
<td style="text-align:left;"><code>category</code></td>
<td>varchar</td>
<td>Description of the business category.</td>
</tr>
</tbody>
</table>
<p>Now let's learn about some of the world's oldest businesses still in operation!</p>

In [1]:
# Import the pandas library under its usual alias 
import pandas as pd

# Load the business.csv file as a DataFrame called businesses
businesses = pd.read_csv('datasets/businesses.csv')
# businesses.head()
# Sort businesses from oldest businesses to youngest
sorted_businesses = businesses.sort_values(by='year_founded')

# Display the first few lines of sorted_businesses
sorted_businesses.head()

business  year_founded category_code country_code
64                    Kongō Gumi           578          CAT6          JPN
94   St. Peter Stifts Kulinarium           803          CAT4          AUT
107        Staffelter Hof Winery           862          CAT9          DEU
106            Monnaie de Paris            864         CAT12          FRA
103               The Royal Mint           886         CAT12          GBR

## 2. The oldest businesses in North America
<p>So far we've learned that Kongō Gumi is the world's oldest continuously operating business, beating out the second oldest business by well over 100 years! It's a little hard to read the country codes, though. Wouldn't it be nice if we had a list of country names to go along with the country codes?</p>
<p>Enter <code>countries.csv</code>, which is also located in the <code>datasets</code> folder. Having useful information in different files is a common problem: for data storage, it's better to keep different types of data separate, but for analysis, we want all the data in one place. To solve this, we'll have to join the two tables together. </p>
<h3 id="countries"><code>countries</code></h3>
<table>
<thead>
<tr>
<th style="text-align:left;">column</th>
<th>type</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;"><code>country_code</code></td>
<td>varchar</td>
<td>ISO 3166-1 3-letter country code.</td>
</tr>
<tr>
<td style="text-align:left;"><code>country</code></td>
<td>varchar</td>
<td>Name of the country.</td>
</tr>
<tr>
<td style="text-align:left;"><code>continent</code></td>
<td>varchar</td>
<td>Name of the continent that the country exists in.</td>
</tr>
</tbody>
</table>
<p>Since <code>countries.csv</code> contains a <code>continent</code> column, merging the datasets will also allow us to look at the oldest business on each continent! </p>

In [3]:
# Load countries.csv to a DataFrame
countries = pd.read_csv("datasets/countries.csv")

# Merge sorted_businesses with countries
businesses_countries = sorted_businesses.merge(countries, on="country_code")

businesses_countries.head()

business  year_founded category_code country_code  \
0                   Kongō Gumi           578          CAT6          JPN   
1  St. Peter Stifts Kulinarium           803          CAT4          AUT   
2        Staffelter Hof Winery           862          CAT9          DEU   
3            Monnaie de Paris            864         CAT12          FRA   
4               The Royal Mint           886         CAT12          GBR   

          country continent  
0           Japan      Asia  
1         Austria    Europe  
2         Germany    Europe  
3          France    Europe  
4  United Kingdom    Europe

## 2(a). Filterning the data (Oldest business of North America)

In [4]:
# Filter businesses_countries to include countries in North America only
north_america = businesses_countries[businesses_countries["continent"] == "North America"]
north_america.head()

business  year_founded category_code country_code  \
22  La Casa de Moneda de México          1534         CAT12          MEX   
28           Shirley Plantation          1638          CAT1          USA   
33         Hudson's Bay Company          1670         CAT17          CAN   
35                Mount Gay Rum          1703          CAT9          BRB   
40                    Rose Hall          1770         CAT19          JAM   

          country      continent  
22         Mexico  North America  
28  United States  North America  
33         Canada  North America  
35       Barbados  North America  
40        Jamaica  North America

## 2(b). Filterning the data (Oldest business of Asian)

In [6]:
asian = businesses_countries[businesses_countries['continent'] == 'Asia']
asian.head()

business  year_founded category_code  \
0                           Kongō Gumi           578          CAT6   
9   Ma Yu Ching's Bucket Chicken House          1153          CAT4   
24                  Çemberlitaş Hamamı          1584         CAT19   
38                         Wadia Group          1736         CAT12   
44                        Pos Malaysia          1800         CAT16   

   country_code   country continent  
0           JPN     Japan      Asia  
9           CHN     China      Asia  
24          TUR    Turkey      Asia  
38          IND     India      Asia  
44          MYS  Malaysia      Asia

## 2(c). Filterning the data (Oldest business of EUROPE)

In [7]:
eropean_old_business = businesses_countries[businesses_countries['continent'] == 'Europe']
eropean_old_business.hea

business  year_founded category_code country_code  \
1       St. Peter Stifts Kulinarium           803          CAT4          AUT   
2             Staffelter Hof Winery           862          CAT9          DEU   
3                 Monnaie de Paris            864         CAT12          FRA   
4                    The Royal Mint           886         CAT12          GBR   
5                        Sean's Bar           900          CAT4          IRL   
6            Marinelli Bell Foundry          1040         CAT12          ITA   
7                  Affligem Brewery          1074          CAT9          BEL   
8                       Munke Mølle          1135         CAT12          DNK   
10                Casa de Ganaderos          1218          CAT1          ESP   
11                  Gasthof Sternen          1230         CAT19          CHE   
12                Bochnia Salt Mine          1248         CAT15          POL   
13             Drohobych salt plant          1250         CAT12          UKR   
14                    Kremnica Mint          1328         CAT12          SVK   
15                The Brand Brewery          1340          CAT9          NLD   
16                  Skyllbergs bruk          1346         CAT12          SWE   
17                  Pivovar Broumov          1348          CAT9          CZE   
18                        Raeapteek          1422         CAT14          EST   
19                  Gostilna Gastuž          1467         CAT19          SVN   
20                           Mousel          1511          CAT9          LUX   
21      CTT-Correios de Portugal SA          1520         CAT16          PRT   
25                        Cēsu Alus          1590          CAT9          LVA   
26                     Piraeus Bank          1606          CAT3          GRC   
29                     Posten Norge          1647         CAT16          NOR   
30                          Fiskars          1649          CAT7          FIN   
32                        Gubernija          1665          CAT9          LTU   
36       Petrodvorets Watch Factory          1721          CAT7          RUS   
37              Kraljevica Shipyard          1729         CAT12          HRV   
39                   Apatin Brewery          1756          CAT9          SRB   
42                    Íslandspóstur          1776         CAT16          ISL   
43                            Zwack          1790          CAT9          HUN   
58                  Pošta Crne Gore          1841         CAT16          MNE   
66   National Bank of Liechtenstein          1861          CAT3          LIE   
68                 Olivaria Brewery          1864          CAT9          BLR   
69                Sarajevska Pivara          1864          CAT9          BIH   
72         Tutunski kombinat Prilep          1873          CAT7          MKD   
75                  Ursus Breweries          1878          CAT9          ROU   
77                       Arsenal AD          1878          CAT8          BGR   
80                  HSBC Bank Malta          1882          CAT3          MLT   
95                   Bank of Cyprus          1899          CAT3          CYP   
106                      ALBtelecom          1912         CAT18          ALB   
119                         Andbank          1930          CAT3          AND   
162            Meridian Corporation          1999         CAT13           XK   

                    country continent  
1                   Austria    Europe  
2                   Germany    Europe  
3                    France    Europe  
4            United Kingdom    Europe  
5                   Ireland    Europe  
6                     Italy    Europe  
7                   Belgium    Europe  
8                   Denmark    Europe  
10                    Spain    Europe  
11              Switzerland    Europe  
12                   Poland    Europe  
13                  Ukraine    Europe  
14                 Slovakia    Europe  
15              Netherland

## 3. The oldest business on each continent
<p>Now we can see that the oldest company in North America is La Casa de Moneda de México, founded in 1534. Why stop there, though, when we could easily find out the oldest business on every continent? </p>

In [ ]:
# Create continent, which lists only the continent and oldest year_founded
continent = businesses_countries.groupby('continent').agg({"year_founded":"min"})

# Merge continent with businesses_countries
merged_continent = continent.merge(businesses_countries, on=['continent', 'year_founded'])

# Subset continent so that only the four columns of interest are included
subset_merged_continent = merged_continent[['continent','country','business','year_founded']]
subset_merged_continent

## 4. Unknown oldest businesses
<p>BusinessFinancing.co.uk wasn't able to determine the oldest business for some countries, and those countries are simply left off of <code>businesses.csv</code> and, by extension, <code>businesses</code>. However, the <code>countries</code> that we created <em>does</em> include all countries in the world, regardless of whether the oldest business is known. </p>
<p>We can compare the two datasets in one DataFrame to find out which countries don't have a known oldest business! </p>

In [ ]:
# Use .merge() to create a DataFrame, all_countries
all_countries = businesses.merge(countries, on="country_code", how="right",  indicator=True)

# Filter to include only countries without oldest businesses
missing_countries = all_countries[all_countries["_merge"] != "both"]

# Create a series of the country names with missing oldest business data
missing_countries_series = missing_countries["country"]

# Display the series
missing_countries_series

## 5. Adding new oldest business data
<p>It looks like we've got some holes in our dataset! Fortunately, we've taken it upon ourselves to improve upon BusinessFinancing.co.uk's work and find oldest businesses in a few of the missing countries. We've stored the newfound oldest businesses in <code>new_businesses</code>, located at <code>"datasets/new_businesses.csv"</code>. It has the exact same structure as our <code>businesses</code> dataset. </p>
<h3 id="new_businesses"><code>new_businesses</code></h3>
<table>
<thead>
<tr>
<th style="text-align:left;">column</th>
<th>type</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;"><code>business</code></td>
<td>varchar</td>
<td>Name of the business.</td>
</tr>
<tr>
<td style="text-align:left;"><code>year_founded</code></td>
<td>int</td>
<td>Year the business was founded.</td>
</tr>
<tr>
<td style="text-align:left;"><code>category_code</code></td>
<td>varchar</td>
<td>Code for the category of the business.</td>
</tr>
<tr>
<td style="text-align:left;"><code>country_code</code></td>
<td>char</td>
<td>ISO 3166-1 3-letter country code.</td>
</tr>
</tbody>
</table>
<p>All we have to do is combine the two so that we've got one more complete list of businesses!</p>

In [ ]:
# Import new_businesses.csv
new_businesses = pd.read_csv('datasets/new_businesses.csv')

# Add the data in new_businesses to the existing businesses
all_businesses = pd.concat([new_businesses, businesses])

# Merge and filter to find countries with missing business data
new_all_countries = all_businesses.merge(countries, on='country_code', how='outer', indicator=True)
new_missing_countries = new_all_countries[new_all_countries['_merge']!='both']

# Group by continent and create a "count_missing" column
count_missing = new_missing_countries.groupby('continent').agg({"country":"count"})
count_missing.columns = ["count_missing"]
count_missing

## 6. The oldest industries
<p>Remember our oldest business in the world, Kongō Gumi? </p>
<table>
<thead>
<tr>
<th style="text-align:right;"></th>
<th style="text-align:left;">business</th>
<th style="text-align:right;">year_founded</th>
<th style="text-align:left;">category_code</th>
<th style="text-align:left;">country_code</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">64</td>
<td style="text-align:left;">Kongō Gumi</td>
<td style="text-align:right;">578</td>
<td style="text-align:left;">CAT6</td>
<td style="text-align:left;">JPN</td>
</tr>
</tbody>
</table>
<p>We know Kongō Gumi was founded in the year 578 in Japan, but it's a little hard to decipher which industry it's in. Information about what the <code>category_code</code> column refers to is in <code>"datasets/categories.csv"</code>: </p>
<h3 id="categories"><code>categories</code></h3>
<table>
<thead>
<tr>
<th style="text-align:left;">column</th>
<th>type</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;"><code>category_code</code></td>
<td>varchar</td>
<td>Code for the category of the business.</td>
</tr>
<tr>
<td style="text-align:left;"><code>category</code></td>
<td>varchar</td>
<td>Description of the business category.</td>
</tr>
</tbody>
</table>
<p>Let's use <code>categories.csv</code> to understand how many oldest businesses are in each category of industry.</p>

In [ ]:
# Import categories.csv and merge to businesses
categories = pd.read_csv("datasets/categories.csv")
businesses_categories = businesses.merge(categories, on="category_code")

# Create a DataFrame which lists the number of oldest businesses in each category
count_business_cats = businesses_categories.groupby("category").agg({"business":"count"})

# Create a DataFrame which lists the cumulative years that businesses from each category have been operating
years_business_cats = businesses_categories.groupby("category").agg({"year_founded":"sum"})

# Rename columns and display the first five rows of both DataFrames
count_business_cats.columns = ["count"]
years_business_cats.columns = ["total_years_in_business"]
display(count_business_cats.head(), years_business_cats.head())

## 7. Restaurant representation
<p>No matter how we measure it, looks like Banking and Finance is an excellent industry to be in if longevity is our goal! Let's zoom in on another industry: cafés, restaurants, and bars. Which restaurants in our dataset have been around since before the year 1800?</p>

In [ ]:
# Filter using .query() for CAT4 businesses founded before 1800; sort results
old_restaurants = businesses_categories.query('year_founded < 1800 and category_code == "CAT4"')

# Sort the DataFrame
old_restaurants = old_restaurants.sort_values("year_founded")
old_restaurants

## 8. Categories and continents
<p>St. Peter Stifts Kulinarium is old enough that the restaurant is believed to have served Mozart - and it would have been over 900 years old even when he was a patron! Let's finish by looking at the oldest business in each category of commerce for each continent. </p>

In [ ]:
# Merge all businesses, countries, and categories together
businesses_categories_countries = businesses_categories.merge(countries, on="country_code")

# Sort businesses_categories_countries from oldest to most recent
businesses_categories_countries = businesses_categories_countries.sort_values("year_founded")

# Create the oldest by continent and category DataFrame
oldest_by_continent_category = businesses_categories_countries.groupby(["continent", "category"]).agg({"year_founded":"min"})
oldest_by_continent_category.head()